### ISSUE TITLE
KILLER FEATURE REQUEST :)

### ISSUE THREAD

**Issue by Plutocratic** [RandomEngy/VidCoder#379]
This is in two parts – the first is easy :-)

1. In the two sizing adjustments (cropping and padding) provide an option to have the pixel values for top/bottom and left/right track one another i.e. only two controls (vertical & horizontal) to tweak instead of four, most of the time. I think it's true to say that most cropping and padding would be done with the **top/bottom values matching** and **left/right values matching.**

2. Setting cropping by number is **tricky and time-consuming**, particularly in a world full of pixels with differing size and shape. What is needed is:

a) **the ability to lock the preview window to a specified aspect ratio.** I'd wager that for 99% of users this will be 16:9 (but you can make other aspect ratios - or the ability to customize - available). Retain the preview window's ability to resize but provide the option to **lock the chosen aspect ratio** (AR) when resizing.

b) The ability to **define cropping points graphically** by the use of a rectangular, mouse-driven box. This "mask box" needs to be moveable in size and position within the source picture but (and this is the important part) needs to have the ability, as with the preview window, to **have its AR locked to one of a number of industry-standard values** such as 16:9, 18:9, 21:9 etc.

The ability to do all necessary cropping etc. visually, without getting concerned with the actual numeric values, will be a GREAT feature.

Oh and finally, to make the above totally useful, you will really need a way of watching any frame of the source in the preview window rather than the limit of 30 arbitrary points currently available.

THANKS :)

**Comment by Ghastor** [comment]
you can manually set aspect ratio as you wish. use external aspect ratio calculator like ARS Calculator.exe to find your aspect ratio into fixed resolution. For example manually set 64:45 for 16/9 pal movie @ 720x576, or 16:15 for 4:3 @ 720x576 pal system.

**Comment by Plutocratic** [comment]
Yes, of course there are ways to do it, but the **simplest and most effective approach** is the visual one I have described!

Just because Napier calculated his logarithms by hand doesn't mean that we should adopt the same approach to calculate aspect ratios. The ideal approach to dealing with oddball aspect ratios is **visual, based on the content** – not merely a number-crunching exercise.

**Comment by RandomEngy** [comment]
I didn't think the mouse would be a great tool for cropping videos because usually you want to be very precise and include all of the picture you can. Does the "auto" cropping feature not do what you want here? Are you using it for more than just cropping out black bars? My approach is usually leave it on Auto to let the video show exactly the content and nothing more. The final aspect ratio might not be "standard" but the player is usually able to deal with that and still shows the picture correctly. And it will play correctly on players with all kinds of aspect ratios.

Is there a reason why you'd need the final picture to exactly match a display aspect ratio? Just trying to get a sense of your use case and why AR-locked cropping is important to you.

**Comment by Plutocratic** [comment]
The main reason for the mouse-driven approach is for dealing with some of the very-wide formats in use i.e. >21:9. I find the massive black bars, top & bottom, disconcerting -- it's like sitting inside a letter box looking out through the slot ;-)

My approach to treating such films is to convert them to something in the region of 18:9 which involves the loss of some picture left and right, but far less intrusive black bars top & bottom. Determination of such a crop is best accomplished visually, with a crop-box that I know to be 18:9 (at the start of the process, even though I might, later, choose to deviate from absolute 18:9). This facility is also useful when cropping, say, material shot on super 16 film which, amongst other things, is designed to have a 16:9 'window' picked out from the frame. Needless to say that this type of task is entirely visual -- you do not even bother to look at the numbers!

The mouse-driven crop (item 2b on my list) is something I fully expect to be a longer-term project, depending on your enthusiasm for the idea ;-) But I reckon that the more important short-term goal ought to be item 2a on my list, the ability to lock the preview window to a known AR. This would enable many of the tasks I described above to be carried out without the benefit of a visual cropping facility, confident in the knowledge that I was previewing a 16:9 screen (and not a 16:10 screen thanks to a careless mouse click). But pro-tem, please implement item 1 on my list ASAP (nice'n'easy) as that alone will make cropping operations far simpler as the preview window will show the effect of adjusting the top and bottom (or left and right) crop simultaneously.

Muchas gracias!

**Comment by RandomEngy** [comment]
The option to match left/right and top/bottom for cropping, I'll add to the backlog. As for a crop locked to a given aspect ratio or a visual mouse-driven crop, I'm not sure how many users like to crop away parts of the picture when transcoding. This might be something I would do if there were more demand, or someone contributed a pull request for it. There are other items on my backlog to tackle that might help more people.

**Comment by Plutocratic** [comment]
OK - fair enough, but could I ask you to seriously prioritize the idea of an option to AR lock the preview window, or at least a numeric means of setting the preview window to a given AR? The problem at the moment is that there is no solid method of setting (or even knowing) the AR of the preview window. The only means we currently have to set the AR of the preview window is by loading a known video "test chart"; it really could do with a more straightforward means of ensuring that the user is viewing a 16:9 (or whatever) sized window.

**Comment by Plutocratic** [comment]
I appreciate the problem - there is some discussion on the question of fixing the AR of a window [here](https://github.com/electron/electron/issues/8036) which may be of use.

But, as a stopgap, could I please ask you to implement an option to lock the top/bottom & left/right values for cropping & padding which really will make this kind of work a lot less frustrating and should be straightforward to implement.

Oh, and while I'm asking stuff, have you ever thought about the desirability of being able to alter the order in which cropping & padding take place? It would be nice to have the option of 
**EITHER** 
cropping first and then padding up to a particular image size
**OR** 
padding the original image and then cropping the overall result

In other words, the option to crop first or pad first.

**Comment by RandomEngy** [comment]
Why would you pad first? Then you'd be cropping off some of the padding you just added. Also the filter order is something that's built into HandBrake and not changeable.

**Comment by Plutocratic** [comment]
Why pad first? If you want to change the target AR (for example, by cropping left and right) while simultaneously adding some padding top and bottom it could be useful in working out the numbers. Sometimes, padding a picture up (say, top and bottom) to a known AR before cropping it left and right can simplify the workflow. Yes, this does entail cropping off some of the padding you just added, but this can be a simpler solution to an otherwise tricky conversion. People are probably doing things with your software that you never imagined or allowed for :) It mainly relates to converting some of the weird stuff created in smartphones, up to a recognisable standard for broadcast.

I thought the padding and cropping was all you (i.e. Vidcoder), as cropping and padding is clearly visible in your preview window before Handbrake has had its mits on the material, so I don't understand how Handbrake's filter order comes into it.

**Comment by RandomEngy** [comment]
For the preview I give my best approximation of what HandBrake will do given the crop and pad settings, but they are the ones doing the actual work when the encode happens.

**Comment by Plutocratic** [comment]
Fascinating and useful to know. I take it, therefore, that padding is done by Handbrake although this feature is not exposed in the Handbrake GUI?

**Comment by RandomEngy** [comment]
Yeah they added it to the core library but they haven't gotten around to exposing it in the UI.
